### ISSUE TITLE
Avoid static type registry lookups in DSL::if_ and DSL::iif calls, if possible

### ISSUE THREAD

**Issue by cies** [jOOQ/jOOQ#17786]
### Your question

We get this warning a lot in our logs:

```
DefaultDataType.getDataType Static type registry : The deprecated static type registry was being accessed for a non-built-in data type: varchar(255) /* models.organization.KycFollowUpStatusRepository$MailType */.

It is strongly recommended not looking up DataType references from Class references by relying on the internal static type registry. For example, avoid calling DSL.val(Object) or DSL.val(Object, Class), and call DSL.val(Object, DataType), providing an explicit DataType ref if your DataType uses a Converter or a Binding. Such a DataType reference can be obtained, for example, using:
- Generated TABLE.COLUMN.getDataType(), if you attached a Converter/Binding to generated code
- From a base type, e.g. SQLDataType.VARCHAR.asConvertedDataType(converterOrBinding)

If you think jOOQ should be able to infer your user type in your particular query, please report a bug here: https://jooq.org/bug See https://github.com/jOOQ/jOOQ/issues/15286 for more details.
```

The code that triggers this error:

```kotlin
    val kfus = KYC_FOLLOW_UP_STATUS.`as`("kfus")
    val od = ORGANIZATION_DETAILS.`as`("od")
    val o = ORGANIZATION.`as`("o")
    DSL
      .select(od.ORGANIZATION_ID, od.VERIFICATION_STATUS)
      .from(od)
      .join(o)
      .on(o.ID.eq(od.ORGANIZATION_ID))
      .where(
        od.ADYEN_VERIFICATION_STATUS.eq(VerificationStatusEnum.NOT_STARTED)
          .and(
            localDateDiff(
              currentLocalDate(),
              o.CREATED_AT.cast(LocalDate::class.java)
            ).eq(DAYS_AFTER_KYC_NOT_STARTED)
          )
      )
      .or(
        od.VERIFICATION_STATUS.ne(VerificationStatusEnum.NOT_STARTED)
          .and(o.SETUP_PERIOD_EXPIRES_AT.lt(setupExpirationWarningThreshold).and(o.SETUP_PERIOD_EXTENDED.isFalse))
      )
      .and(
        notExists(
          select(kfus.ORGANIZATION_ID).from(kfus).where(kfus.ORGANIZATION_ID.eq(od.ORGANIZATION_ID)).and(
            kfus.TYPE.eq(
              if_(
                od.VERIFICATION_STATUS.eq(VerificationStatusEnum.NOT_STARTED),
                MailType.NOT_STARTED,
                MailType.NOT_COMPLETED
              )
            )
          )
        )
      )
      .and(o.DELETED.isFalse)
      .and(o.DELETE_ON.isNull)
      .fetch { // <--- this line triggers the error according to the stack trace
        attachedRecord(
          KycFollowUpStatusRecord(
            organizationId = it.value1()!!,
            type = if (it.value2() == VerificationStatusEnum.NOT_STARTED) MailType.NOT_STARTED else MailType.NOT_COMPLETED
          )
        )
      }
```

There's also a stack trace part of the warning:

```
org.jooq.impl.DefaultRenderContext.visit0(DefaultRenderContext.java:731)
org.jooq.impl.AbstractContext.visit(AbstractContext.java:362)
org.jooq.impl.AbstractContext.visit(AbstractContext.java:261)
org.jooq.impl.SelectQueryImpl.toSQLReference0(SelectQueryImpl.java:2636)
org.jooq.impl.SelectQueryImpl.lambda$toSQLReferenceLimitDefault$10(SelectQueryImpl.java:2101)
org.jooq.impl.AbstractContext.data(AbstractContext.java:419)
org.jooq.impl.SelectQueryImpl.toSQLReferenceLimitDefault(SelectQueryImpl.java:2101)
org.jooq.impl.SelectQueryImpl.accept0(SelectQueryImpl.java:1965)
org.jooq.impl.SelectQueryImpl.accept(SelectQueryImpl.java:1598)
org.jooq.impl.DefaultRenderContext.visit0(DefaultRenderContext.java:731)
org.jooq.impl.AbstractContext.visit(AbstractContext.java:362)
org.jooq.impl.AbstractDelegatingQuery.accept(AbstractDelegatingQuery.java:90)
org.jooq.impl.DefaultRenderContext.visit0(DefaultRenderContext.java:731)
org.jooq.impl.AbstractContext.visit(AbstractContext.java:362)
org.jooq.impl.Tools.visitSubquery(Tools.java:3048)
org.jooq.impl.Tools.visitSubquery(Tools.java:3021)
org.jooq.impl.Exists.accept(Exists.java:116)
org.jooq.impl.DefaultRenderContext.visit0(DefaultRenderContext.java:731)
org.jooq.impl.AbstractContext.visit(AbstractContext.java:362)
org.jooq.impl.AbstractContext.visit(AbstractContext.java:261)
org.jooq.impl.Not.accept(Not.java:111)
org.jooq.impl.DefaultRenderContext.visit0(DefaultRenderContext.java:731)
org.jooq.impl.AbstractContext.visit(AbstractContext.java:362)
org.jooq.impl.Expression.acceptAssociative(Expression.java:939)
org.jooq.impl.And.accept(And.java:121)
org.jooq.impl.DefaultRenderContext.visit0(DefaultRenderContext.java:731)
org.jooq.impl.AbstractContext.visit(AbstractContext.java:362)
org.jooq.impl.AbstractContext.visit(AbstractContext.java:261)
org.jooq.impl.ConditionProviderImpl.accept(ConditionProviderImpl.java:152)
org.jooq.impl.DefaultRenderContext.visit0(DefaultRenderContext.java:731)
org.jooq.impl.AbstractContext.visit(AbstractContext.java:362)
org.jooq.impl.AbstractContext.visit(AbstractContext.java:261)
org.jooq.impl.SelectQueryImpl.toSQLReference0(SelectQueryImpl.java:2636)
org.jooq.impl.SelectQueryImpl.lambda$toSQLReferenceLimitDefault$10(SelectQueryImpl.java:2101)
org.jooq.impl.AbstractContext.data(AbstractContext.java:419)
org.jooq.impl.SelectQueryImpl.toSQLReferenceLimitDefault(SelectQueryImpl.java:2101)
org.jooq.impl.SelectQueryImpl.accept0(SelectQueryImpl.java:1965)
org.jooq.impl.SelectQueryImpl.accept(SelectQueryImpl.java:1598)
org.jooq.impl.DefaultRenderContext.visit0(DefaultRenderContext.java:731)
org.jooq.impl.AbstractContext.visit(AbstractContext.java:362)
org.jooq.impl.DefaultRenderContext$Rendered.rendered(DefaultRenderContext.java:1082)
org.jooq.impl.AbstractQuery.getSQL0(AbstractQuery.java:498)
org.jooq.impl.AbstractQuery.execute(AbstractQuery.java:303)
org.jooq.impl.AbstractResultQuery.fetchLazy(AbstractResultQuery.java:301)
org.jooq.impl.AbstractResultQuery.fetchLazyNonAutoClosing(AbstractResultQuery.java:322)
org.jooq.impl.SelectImpl.fetchLazyNonAutoClosing(SelectImpl.java:3256)
org.jooq.impl.ResultQueryTrait.collect(ResultQueryTrait.java:360)
org.jooq.impl.ResultQueryTrait.fetch(ResultQueryTrait.java:1465)
our.code.package.ClassName.methodName(OurFile.kt:<see comment in code pointing to the line>)
```

Based on the warning i have no clue what we're doing wrong here. How am I "looking up DataType references from Class references by relying on the internal static type registry"? I dont see our code "calling DSL.val(Object)" or "DSL.val(Object, Class)"... Hence I raise this issue in hope to get some pointer to what is going wrong here, and how to fix it.

Thanks for looking into it.

### jOOQ Version

3.19.15

### Database product and version

MySQL 8.0.33

### Java Version

JDK21, Kotlin 2.0.20

### JDBC / R2DBC driver name and version (include name if unofficial driver)

mysql-connector-j 9.0.0

**Comment by lukaseder** [comment]
Thanks a lot for your report. This expression is probably causing it:

```java
kfus.TYPE.eq(
  if_(
    od.VERIFICATION_STATUS.eq(VerificationStatusEnum.NOT_STARTED),
    MailType.NOT_STARTED,
    MailType.NOT_COMPLETED
  )
)
```

It's the same problem as with `CASE` expressions:

- https://github.com/jOOQ/jOOQ/issues/16750

We should fix this on our side. You can either ignore this particular warning, or provide explicit data types:

```java
kfus.TYPE.eq(
  if_(
    od.VERIFICATION_STATUS.eq(VerificationStatusEnum.NOT_STARTED),
    val(MailType.NOT_STARTED, kfus.TYPE),
    val(MailType.NOT_COMPLETED, kfus.TYPE)
  )
)
```

Can you confirm this removes the warning?

**Comment by cies** [comment]
> Can you confirm this removes the warning?

Just checked to confirm it: it works.

It's a bit hard though to find out all places it goes wrong, as the warnings are triggered at runtime.

**Comment by lukaseder** [comment]
Fixed for jOOQ 3.20.

Just like other, comparable changes, this change won't be backported as it might subtly change behaviour of existing queries when the `Converter` works on built-in data types, e.g. `Converter<Integer, Integer>` or `Converter<Integer, String>`. For example, with `if_(condition, field, 1)`, the value `1` will now be converted instead of being bound as it is.

**Comment by cies** [comment]
@lukaseder 

> Fixed for jOOQ 3.20.

Yay! Thanks for that.

As promised I will now report the other places we had to "fix" to prevent the warnings. There was one, we changed:

```kotlin
           .where(
              ac2.filterOrganization(organization.id!!)
                .and(ac2.TYPE.eq(adyenCapabilityRecord.type))
            )
```

to:

```kotlin
              .where(ac2.filterOrganization(organization.id!!)
                // The following `DSL.val` calls can be removed if jOOQ fixes https://github.com/jOOQ/jOOQ/issues/17786
                .and(ac2.TYPE.eq(DSL.`val`(adyenCapabilityRecord.type, ac1.TYPE)))
              )
```

It's currently running in production w/o warnings of this kind. If we stumble on warnings, I will report them here!

**Comment by lukaseder** [comment]
> It's currently running in production w/o warnings of this kind. If we stumble on warnings, I will report them here!

Great to hear. Ideally with new issues, yes!
